src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+    std::list<std::shared_ptr<LedgerFile>> files;
+
+    // Cache of ledger files for reading
+    size_t max_read_cache_size;
+    std::list<std::shared_ptr<LedgerFile>> files_read_cache;
+
+    const size_t chunk_threshold;
+    size_t last_idx = 0;
+    size_t compacted_idx = 0;
+
+    // True if a new file should be created when writing an entry
+    bool require_new_file;
+
+    auto get_it_contains_idx(size_t idx) const
+    {
+      if (idx == 0)
+      {
+        return files.end();
+      }
+
+      auto f = std::upper_bound(
+        files.begin(),
+        files.end(),
+        idx,
+        [](size_t idx, const std::shared_ptr<LedgerFile>& f) {
+          return (idx <= f->get_last_idx());
+        });
+
+      return f;
+    }
+
+    std::shared_ptr<LedgerFile> get_file_from_cache(size_t idx)
+    {
+      if (idx == 0)
+      {
+        return nullptr;
+      }
+
+      // First, try to find file from read cache
+      for (auto const& f : files_read_cache)
+      {
+        if (f->get_start_idx() <= idx && idx <= f->get_last_idx())
+        {
+          return f;
+        }
+      }
+
+      // If the file is not in the cache, find the file from the ledger
+      // directory
+      std::map<size_t, std::string> all_files;
+      for (auto const& f : fs::directory_iterator(ledger_dir))
+      {
+        all_files.emplace(
+          get_start_idx_from_file_name(f.path().filename()),
+          f.path().filename());
+      }
+
+      auto f_ = std::upper_bound(
+        all_files.rbegin(),
+        all_files.rend(),
+        idx,
+        [](size_t idx, const std::pair<size_t, std::string>& v) {
+          return (idx >= v.first);
+        });
+
+      if (f_ == all_files.rend())
+      {
+        LOG_FAIL_FMT("Could not find ledger file on disk for idx {}", idx);
+        return nullptr;
+      }
+
+      // Emplace file in the max-sized read cache
+      auto match_file = std::make_shared<LedgerFile>(ledger_dir, f_->second);
+      if (files_read_cache.size() >= max_read_cache_size)
+      {
+        files_read_cache.erase(files_read_cache.begin());
+      }
+      files_read_cache.emplace_back(match_file);
+
+      return match_file;
+    }
+
+    std::shared_ptr<LedgerFile> get_file_from_idx(size_t idx)
+    {
+      if (idx == 0)
+      {
+        return nullptr;
+      }
+
+      // First, check if the file is in the list of files open for writing
+      auto f = std::upper_bound(
+        files.rbegin(),
+        files.rend(),
+        idx,
+        [](size_t idx, const std::shared_ptr<LedgerFile>& f) {
+          return idx >= f->get_start_idx();
+        });
+
+      if (f != files.rend())
+      {
+        return *f;
+      }
+
+      // Otherwise, return file from read cache
+      return get_file_from_cache(idx);
+    }
+
+    std::shared_ptr<LedgerFile> get_latest_file() const
+    {
+      if (files.empty())
+      {
+        return nullptr;
+      }
+      return *(files.rbegin()++);
```

Comment:
This looks like it returns the last-but-one file? `files.rbegin()` should be the last element in files (its not the same as `files.end()`)

Reply:
This somehow returns the last element in the list. But `files.back()` should be preferred.